x-pack/libbeat/management/managerV2.go

```diff
@@ -251,38 +315,29 @@ func (cm *BeatV2Manager) unitListen() {
 				cm.logger.Debug("Received sighup, stopping")
 			}
 			cm.isRunning = false
-			unit, mainExists := cm.getMainUnit()
-			if mainExists {
-				_ = unit.UpdateState(client.UnitStateStopping, "stopping beat", nil)
-			}
-			cm.client.Stop()
+			cm.UpdateStatus(lbmanagement.Stopping, "Stopping")
 			return
 		case change := <-cm.client.UnitChanges():
 			switch change.Type {
 			// Within the context of how we send config to beats, I'm not sure there is a difference between
 			// A unit add and a unit change, since either way we can't do much more than call the reloader
 			case client.UnitChangedAdded:
-				// At this point we also get a log level, however I'm not sure the beats core logger provides a
-				// clean way to "just" change the log level, without resetting the whole log config
-				state, _, _ := change.Unit.Expected()
-				cm.logger.Debugf("Got unit added: %s, type: %s expected state: %s", change.Unit.ID(), change.Unit.Type(), state.String())
-				go cm.handleUnitReload(change.Unit)
-
+				cm.addUnit(change.Unit)
+				cm.triggerReload()
 			case client.UnitChangedModified:
-				state, _, _ := change.Unit.Expected()
-				cm.logger.Debugf("Got unit modified: %s, type: %s expected state: %s", change.Unit.ID(), change.Unit.Type(), state.String())
-				// I'm assuming that a state STOPPED just tells us to shut down the entire beat,
-				// as such we don't really care about updating via a particular unit
-				if state == client.UnitStateStopped {
-					cm.stopBeat()
-				} else {
-					go cm.handleUnitReload(change.Unit)
-				}
-
+				cm.modifyUnit(change.Unit)
+				cm.triggerReload()
 			case client.UnitChangedRemoved:
-				cm.logger.Debugf("Got unit removed: %s", change.Unit.ID())
 				cm.deleteUnit(change.Unit)
 			}
+		case <-cm.reloadCh:
```

Comment:
I'm not sure why there's so much logic involved in the reloading process? `triggerReload()` and `reload()` are both only called in this for block, it might just be simpler to call `reload()` in a goroutine or something? Also, why is `reload()` getting its own map of the units and not just using `cm.units`?

Reply:
This is because performing `reload` might result in the config logic of the beat to call `UpdateStatus` which will then grab the mutex for the units.

So to ensure that a dead lock doesn't occur we ensure that the actual reload logic is performed in the main loop of the manager, but not in a path that would hold that mutex. So that is why a copy is sent to the reload function. Each `client.Unit` has an internal lock for state so it is also save to have a pointer to the same unit.

Reply:
When you mention a deadlock starting with `UpdateStatus` I assume you mean the `Reload()` call blocking while the beat waits for `UpdateStatus`? 

I wonder if we can do something like have a `addUnit()` return a new copy of the unit map that's sent to `reload()` so we can avoid the extra `reloadCh`? If not, can you at least add a comment explaining the loop, was a tad confusing at first.

Reply:
Can you add a comment explaining the potential for deadlock here so it's obvious the next time we read this code?

Reply:
I have added a comment as well as add some debounce to consolidate unit changes into a single reload. It makes it simpler to understand as well.